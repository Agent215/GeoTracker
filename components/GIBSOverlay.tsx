/**
 * Custom component to build an array of strings that range over a date and then return a urlTile object to be loaded
 * into the mapview. The intent is to change urlTiles using an external counter tied to setTimeout at approximately 1
 * second per tile.
 */

import { UrlTile } from "react-native-maps";
import React, { useState } from "react"



const GIBSOverlay = (props) => {
    //const [category, setCategory] = useState('')
    //let date = props.startDate;
    //let endDate = props.endDate;
    //let count = props.count
    //let isPlaying = props.isPlaying;

    //let gibsUrlTemplate = urlBuilder(setCategory(props.category), date, endDate);
    const [count, setCount] = useState(0)
    let tempUrl = urlBuilder(props.category, props.date)
    while(props.playing){
        setTimeout(() => {
            setCount(count+1)
        }, 500)
    }
    let gibsTile = (<UrlTile 
    urlTemplate={tempUrl}
    maximumZ={19}
    flipY={false}
    //key={Date.now()}
    />)

    return (gibsTile);
}

function urlBuilder(category, date){
    let urlStart = 'https://gibs.earthdata.nasa.gov/wmts/epsg3857/best/';                        //url being built
    //let urlArray = [];              //empty array to be filled with strings for urlTemplate
    let layer = category;             //category of url
    //let firstDate = dateStart;      //first date to build urls from
    //let lastDate = dateEnd;         //last date of urls
    let urlDate = date;          //date of url string for testing

    //while loop to build url strings for building urlTiles for GIBSOverlay set in date range
    //while(firstDate != lastDate/*TODO: function from date-fns to add a day */){
        let tempUrl = urlStart;     //string to hold url string being built

        //handle layer of url using matching GIBS category string
        if (layer == undefined) {
            tempUrl = '';
            console.log('Category undefined.')
            return tempUrl;
        }
        if (layer == ""){
            tempUrl = '';
            return tempUrl;
        }
        if (layer == "clouds") tempUrl = tempUrl + 'IMERG_Precipitation_Rate/default/';
        if (layer == "temp") tempUrl = tempUrl + 'AIRS_L3_Surface_Air_Temperature_Monthly_Night/default/';
        if (layer == "precipitation") tempUrl = tempUrl + 'IMERG_Precipitation_Rate/default/';
        if (layer == "wind") tempUrl = tempUrl + 'Aquarius_Wind_Speed_L3_7Day_Snapshot/default/';
        if (layer == "pressure") tempUrl = tempUrl + 'MODIS_Terra_Cloud_Top_Pressure_Day/default/';
        if (layer == "none"){
            tempUrl = '';
            return tempUrl;
        }
        //handle date of url
        tempUrl = tempUrl + date;      //TODO: add date-fns and parse then format date from provided
        /*uncomment if tileMatrixSet is used later//handle tile matrix set to be associated with zoom level
        if(layer == 'clouds') tempUrl = tempUrl + '';
        if(layer == 'temp') tempUrl = tempUrl + '';
        if(layer == 'precipitation') tempUrl = tempUrl + '';
        if(layer == 'wind') tempUrl = tempUrl + '';
        if(layer == 'pressure') tempUrl = tempUrl + '';
        */
        //remove this if tile matrix set is not all the same thing
        tempUrl = tempUrl + '/GoogleMapsCompatible_Level6/';
        //add tail of url template
        tempUrl = tempUrl + '{z}/{y}/{x}.png';
        console.log("Url is: " + tempUrl)
        //add temp url to end of url array
        //urlArray.push(tempUrl)
        //console.log("Length of urlArry: " + urlArray.length);
    //}
    //return urlArray;
    return tempUrl;
}

export default GIBSOverlay;