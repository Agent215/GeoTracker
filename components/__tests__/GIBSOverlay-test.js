import * as React from 'react';
import renderer from 'react-test-renderer';
import GIBSOverlay from '../../components/GIBSOverlay';
import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

enzyme.configure({ adapter: new Adapter() });

//array of categories to test randomly
const categories = ["clouds", "temp", "precipitation", "wind", "pressure"]

describe('<GIBSOverlay />', () => {

    it(`GIBSOverlay renders Correctly on Load`, () => {

        const GIBSOverlayComponent = renderer.create(<GIBSOverlay />).toJSON()    
        expect(GIBSOverlayComponent).toMatchSnapshot();
        });

    it('GIBSOverlay is null if no weather to be displayed', () => {

        const GIBSOverlayComponent = renderer.create(<GIBSOverlay category={"none"} gibsVisible={true} />).toJSON();
    })

    it(`GIBSOverlay is null if weather to be displayed`, () => {

        const GIBSOverlayComponent = renderer.create(<GIBSOverlay category={ categories[Math.floor(Math.random() * categories.length)] } gibsVisible={false} />).toJSON()   
        expect(GIBSOverlayComponent).toBeNull();
        });

    it('GIBSOverlay is not null if gibs is visible', () => {

        const GIBSOverlayComponent = renderer.create(<GIBSOverlay category={"clouds"} gibsVisible={true} />).toJSON()
        expect(GIBSOverlayComponent).not.toBeNull();
    })
})