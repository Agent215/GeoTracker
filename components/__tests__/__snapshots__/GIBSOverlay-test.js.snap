// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GIBSOverlay /> GIBSOverlay renders Correctly on Load 1`] = `null`;
